client/rabbitmq.go

```diff
@@ -97,91 +133,90 @@ func (c *RabbitClient) Call(ctx context.Context, serviceName, endpoint string, r
 
 	// Don't even try to send if not connected
 	if !c.connection.IsConnected() {
-		return errors.Wrap(fmt.Errorf("Not connected to AMQP"))
+		return nil, errors.Wrap(fmt.Errorf("Not connected to AMQP"))
 	}
 
-	routingKey := c.buildRoutingKey(serviceName, endpoint)
+	replyChannel := c.inflight.push(req.Id())
 
-	correlation, err := uuid.NewV4()
-	if err != nil {
-		log.Errorf("[Client] Failed to create unique request id: %v", err)
-		return errors.Wrap(err) // @todo custom error code
-	}
-
-	log.Debugf("[Client] Dispatching request to %s with correlation ID %s", routingKey, correlation.String())
-
-	replyChannel := c.inflight.push(correlation.String())
-
-	requestBody, err := proto.Marshal(req)
-	if err != nil {
-		log.Errorf("[Client] Failed to marshal request: %v", err)
-		return errors.Wrap(err) // @todo custom error code
-	}
+	routingKey := buildRoutingKey(req.Service(), req.Endpoint())
+	log.Debugf("[Client] Dispatching request to %s with correlation ID %s", routingKey, req.Id())
 
+	// Build message from request
 	message := amqp.Publishing{
-		CorrelationId: correlation.String(),
+		CorrelationId: req.Id(),
 		Timestamp:     time.Now().UTC(),
-		Body:          requestBody,
+		Body:          req.Payload(),
 		ReplyTo:       c.replyTo,
+		Headers: amqp.Table{
+			"Content-Type":     req.ContentType(),
+			"Content-Encoding": "request",
```

Comment:
Yes all the responses are sent back with either `response` or `error`, thought it made sense to standardise this with a `request` type. Also if we add heartbeart/pings these could use this field to identify themselves.